src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1140,6 +1230,123 @@ private void IssueBuildSubmissionToScheduler(BuildSubmission submission, bool al
+                }
+
+                // TODO: Handle multiple entry points
+                var entryNode = projectGraph.EntryPointNodes.First();
+                var targetLists = projectGraph.GetTargetLists(submission.BuildRequestData.TargetNames);
+
+                var waitHandle = new AutoResetEvent(true);
+                var graphBuildStateLock = new object();
+
+                var blockedNodes = new HashSet<ProjectGraphNode>(projectGraph.ProjectNodes);
+                var finishedNodes = new HashSet<ProjectGraphNode>(projectGraph.ProjectNodes.Count);
+                var buildingNodes = new Dictionary<BuildSubmission, ProjectGraphNode>();
+                while (blockedNodes.Count > 0 || buildingNodes.Count > 0)
+                {
+                    waitHandle.WaitOne();
+
+                    lock (graphBuildStateLock)
+                    {
+                        var unblockedNodes = blockedNodes
+                            .Where(node => node.ProjectReferences.All(projectReference => finishedNodes.Contains(projectReference)))
+                            .ToList();
+                        foreach (var node in unblockedNodes)
+                        {
+                            var targetList = targetLists[node];
+                            if (targetList.Count == 0)
+                            {
+                                // An empty target list here means "no targets" instead of "default targets", so don't even build it.
+                                finishedNodes.Add(node);
+                                blockedNodes.Remove(node);
+
+                                waitHandle.Set();
+
+                                continue;
+                            }
+
+                            var request = new BuildRequestData(node.ProjectInstance, targetList.ToArray());
+
+                            // TODO Tack onto the existing submission instead of pending a whole new submission for every node
+                            // Among other things, this makes BuildParameters.DetailedSummary produce a summary for each node, which is not desirable.
+                            var innerBuildSubmission = PendBuildRequest(request);
+                            buildingNodes.Add(innerBuildSubmission, node);
+                            blockedNodes.Remove(node);
+                            innerBuildSubmission.ExecuteAsync(finishedBuildSubmission =>
+                            {
+                                ProjectGraphNode finishedNode;
+                                lock (graphBuildStateLock)
+                                {
+                                    finishedNode = buildingNodes[finishedBuildSubmission];
+
+                                    finishedNodes.Add(finishedNode);
+                                    buildingNodes.Remove(finishedBuildSubmission);
+                                }
+
+                                if (finishedNode == entryNode)
```

Comment:
There could be multiple entrypoints, and an entrypoint can actually not be a root. Shouldn't the results be reported only when all the ProjectGraph.GraphRoots are done? Or simpler, when the finishedNodes contains all the nodes in the graph.